lib/reversetunnel/localsite.go

```diff
@@ -281,38 +281,45 @@ func (s *localSite) getConn(params DialParams) (conn net.Conn, useTunnel bool, e
 	if params.To != nil {
 		dreq.Address = params.To.String()
 	}
+
 	// If server ID matches a node that has self registered itself over the tunnel,
-	// return a connection to that node. Otherwise net.Dial to the target host.
-	conn, err = s.dialTunnel(dreq)
-	if err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, false, trace.Wrap(err)
-		}
+	// return a tunnel connection to that node. Otherwise net.Dial to the target host.
+	conn, tunnelErr := s.dialTunnel(dreq)
+	if tunnelErr == nil {
+		return conn, true, nil
+	}
 
-		// Connections to applications should never occur over a direct dial, return right away.
-		if params.ConnType == services.AppTunnel {
-			return nil, false, trace.ConnectionProblem(err, "failed to connect to application")
-		}
+	if !trace.IsNotFound(tunnelErr) {
+		return nil, false, trace.Wrap(tunnelErr)
+	}
 
-		// This node can only be reached over a tunnel, don't attempt to dial
-		// remotely.
-		if params.To.String() == "" {
-			return nil, false, trace.ConnectionProblem(err, "node is offline, please try again later")
-		}
-		tunnelErr := trace.Errorf("dialing through a tunnel: %v", err)
-		// If no tunnel connection was found, dial to the target host.
-		dialer := proxy.DialerFromEnvironment(params.To.String())
-		conn, err = dialer.DialTimeout(params.To.Network(), params.To.String(), defaults.DefaultDialTimeout)
-		if err != nil {
-			return nil, false, trace.NewAggregate(tunnelErr, trace.Errorf("dialing directly: %v", err))
-		}
+	s.log.WithError(tunnelErr).Debug("Error occurred while dialing through a tunnel.")
```

Comment:
Nit: At this point the only possible outcome is that the tunnel hasn't been found, right? So I'd just log that, `tunnelErr` is always gonna be trace.NotFound here. Also, mention which particular tunnel this is for.

```go
s.log.Debugf("Tunnel %v not found.", XXX)
```

Reply:
I'd prefer to keep it as is, the received error object may be annotated with additional fields or a more detailed message, we probably don't need to drop such information.

There is already one such additional field (cluster name):

```
DEBU [PROXY:SER] Error occurred while dialing through a tunnel. address:127.0.0.1:3022 error:no tunnel connection found: no node reverse tunnel for 3440ba8f-2a0c-4e88-90b1-202703c91bad.graviton-auth found trace.fields:map[cluster:graviton-auth] reversetunnel/localsite.go:299
```

Reply:
Ah, in that case it does make sense to keep it, yes.